### PR TITLE
Glass brain hemispheres

### PR DESCRIPTION
**Pull request by MartinPerez**
Glass brain hemispheres implementation to address #942.

tests, doc update and examples update included.

Here result images from the plot_demo_glass_brain_extensive example.

figure_6
figure_7
figure_8
figure_9
figure_10

### CODE REVIEW

**Comment by AlexandreAbraham** [comment]
This looks really nice!

**Comment by MartinPerez** [comment]
sorry rebasing ...

**Comment by GaelVaroquaux** [comment]
This does look very nice! On the images at the top (with a white background), there is a strange yellow background color.

**Comment by MartinPerez** [comment]
You mean the two first images? I see it white :S

**Comment by GaelVaroquaux** [comment]
I am positive that there is something in the background. I can see the rectangles.

**Comment by MartinPerez** [comment]
in two screens (desktop, laptop) I see it white. No yellow background... @AlexandreAbraham what do you see?

**Comment by AlexandreAbraham** [comment]
There is a yellow background. Color 255, 255, 251. Gimp can help:

yellow

File: examples/01_plotting/plot_demo_glass_brain.py [hunk 1]
```diff
@@ -20,10 +20,19 @@
...
 
 plotting.plot_glass_brain(
     localizer_tmap_filename, title='plot_glass_brain',
     black_bg=True, display_mode='xz', threshold=3)
 
+# Hemispheric saggital cuts
+plotting.plot_glass_brain(
+    localizer_tmap_filename, display_mode='lyrz', threshold=3)
+
+plotting.plot_glass_brain(
```
Comment (GaelVaroquaux) [inline examples/01_plotting/plot_demo_glass_brain.py]: We want to keep this example as simple as possible. I suggest that we remove that last line, which belongs more to the advanced example.

File: examples/01_plotting/plot_demo_glass_brain_extensive.py [hunk 2]
```diff
@@ -35,4 +36,20 @@
 plotting.plot_glass_brain(localizer_tmap_filename, threshold=3,
                           colorbar=True, plot_abs=False)
 
+# Hemispheric saggital cuts
+plotting.plot_glass_brain(localizer_tmap_filename, threshold=3,
+                          display_mode='lyrz')
```
Comment (GaelVaroquaux) [inline examples/01_plotting/plot_demo_glass_brain_extensive.py]: Can we have a demo of "lzr", which I suspect is the mode useful mode.

File: nilearn/plotting/displays.py [hunk 3]
```diff
@@ -350,6 +350,120 @@ def _add_lines(self, line_coords, line_values, cmap,
...
+                               extent=(xmin, xmax, zmin, zmax),
+                               **kwargs)
+
+        self.add_object_bounds((xmin_, xmax_, zmin_, zmax_))
+
+        return im
+
+    def draw_left_right(self, size, bg_color, **kwargs):
+        if self.direction in 'lr':
+            return
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: Am I wrong to think that if we modify the base class "draw_left_right" method, around line 115, we don't need to reimplement this method in the derived class?

File: nilearn/plotting/displays.py [hunk 4]
```diff
@@ -350,6 +350,120 @@ def _add_lines(self, line_coords, line_values, cmap,
...
+
+        ax.text(.9, .95, 'R',
+                transform=ax.transAxes,
+                horizontalalignment='right',
+                verticalalignment='top',
+                size=size,
+                bbox=dict(boxstyle="square,pad=0", ec=bg_color, fc=bg_color),
+                **kwargs)
+
+    def transform_to_2d(self, data, affine):
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: It seems to me that there's a lot of copy/paste with regards to the base-class method (in GlassBrainAxes). I would prefer modifying it to support l/r projection.

File: nilearn/plotting/displays.py [hunk 5]
```diff
@@ -1312,13 +1481,35 @@ class YZProjector(OrthoProjector):
...
+
+class LProjector(OrthoHemisphericProjector):
+    _cut_displayed = 'l'
+    _default_figsize = [2.6, 2.3]
+
+
+class RProjector(OrthoHemisphericProjector):
+    _cut_displayed = 'r'
+    _default_figsize = [2.6, 2.3]
+
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: I'd like an "lzr" mode, I think.

File: nilearn/plotting/img_plotting.py [hunk 6]
```diff
@@ -963,14 +963,16 @@ def plot_glass_brain(stat_map_img,
             See http://nilearn.github.io/manipulating_visualizing/manipulating_images.html#niimg.
             The statistical map image. It needs to be in MNI space
             in order to align with the brain schematics.
-        output_file : string, or None, optional
+        output_file : string, or None, optional. 'ortho' by default.
```
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: Is this change intentional? I have the feeling that it is not right.

File: nilearn/plotting/tests/test_displays.py [hunk 7]
```diff
@@ -40,6 +41,16 @@ def test_demo_ortho_projector():
...
 
+def test_demo_ortho_hemispheric_projector():
+    # This is only a smoke test
+    img = load_mni152_template()
+    oprojector = OrthoHemisphericProjector.init_with_figure(img=img)
+    oprojector.add_overlay(img, cmap=plt.cm.gray)
+    with tempfile.TemporaryFile() as fp:
+        oprojector.savefig(fp)
+    oprojector.close()
+
```
Comment (GaelVaroquaux) [inline nilearn/plotting/tests/test_displays.py]: Could we also test the functionality with a call to plot_glass_brain.

**Comment by MartinPerez** [comment]
I fixed the background issue and discovered on the way that I have some yellow daltonism, haha.

figure_10

File: examples/01_plotting/plot_demo_glass_brain_extensive.py (hunk 2 above)
Comment (MartinPerez) [inline examples/01_plotting/plot_demo_glass_brain_extensive.py]: no problem

File: examples/01_plotting/plot_demo_glass_brain.py (hunk 1 above)
Comment (MartinPerez) [inline examples/01_plotting/plot_demo_glass_brain.py]: ok

File: nilearn/plotting/img_plotting.py (hunk 6 above)
Comment (MartinPerez) [inline nilearn/plotting/img_plotting.py]: I just mention the default value already implemented. You mean I should not mention it, or that the default itself is wrong?

File: nilearn/plotting/tests/test_displays.py (hunk 7 above)
Comment (MartinPerez) [inline nilearn/plotting/tests/test_displays.py]: Like in the examples?

File: nilearn/plotting/img_plotting.py (hunk 6 above)
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: > I just mention the default value already implemented. You mean I should not
> mention it, or that the default itself is wrong?

The default is "None": we are talking about "output_file", aren't we?

File: nilearn/plotting/displays.py (hunk 4 above)
Comment (MartinPerez) [inline nilearn/plotting/displays.py]: I also noticed that some changes would just mean one line in the parent classes. Nonetheless I put all the changes in the GlassBrainHemisphericAxes class for 2 reasons. First its not just GlassBrainAxes but also BaseAxes that would have modifications, I thought having spread changes across multiple classes for one feature might be confusing in the future and could break some other derived class (so far its assume for all the plotting framework that only xyz exist). Second I am not sure how the changes introduced would interact with the connectomics plots that also use the GlassBrainAxes class. I think putting this together and checking how this should be used with the connectomics should be a different PR. @GaelVaroquaux what do you think?

File: nilearn/plotting/tests/test_displays.py (hunk 7 above)
Comment (GaelVaroquaux) [inline nilearn/plotting/tests/test_displays.py]: Yes, with a simple smoketest.

File: nilearn/plotting/img_plotting.py (hunk 6 above)
Comment (MartinPerez) [inline nilearn/plotting/img_plotting.py]: sorry you are right. mistake during the rebase merge just done.

**Comment by MartinPerez** [comment]
Addressed all comments. Added lyr, lzr and lzry and actually generalized the code to accept any order of labels in the future, which also simplified the hemispheric projector class. Added smoke test on test_img_plotting.py.

Examples here:

figure_11
figure_12

**Comment by chrisfilo** [comment]
Excellent addition! LGTM.

**Comment by MartinPerez** [comment]
circle ci seems to fail due to sphinx and doc generation? Can anyone please give me a hand to understand this error. Codacy fails for what was pointed out by @GaelVaroquaux  about code duplication in displays.py in GlassBrainHemisphericAxes class

File: nilearn/plotting/img_plotting.py [hunk 8]
```diff
@@ -963,14 +963,16 @@ def plot_glass_brain(stat_map_img,
...
             The statistical map image. It needs to be in MNI space
             in order to align with the brain schematics.
-        output_file : string, or None, optional
+        output_file : string, or None, optional.
             The name of an image file to export the plot to. Valid extensions
             are .png, .pdf, .svg. If output_file is not None, the plot
             is saved to a file, and the display is closed.
-        display_mode: {'ortho', 'x', 'y', 'z', 'xy', 'xz', 'yz'}
+        display_mode : {'ortho', 'x', 'y', 'z', 'xz', 'yx', 'yz', 'l', 'r',
+                        'lr', 'lzr', 'lyr', 'lzry', 'lyrz'}. Default is 'ortho'.
```
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: Your problem with circleCI might be that you cannot have a line return here. It's possible that this just needs to be a very long line.

**Comment by MartinPerez** [comment]
circle ci pass now. The appveyor new error seems unrelated to the changes of this PR.

======================================================================
ERROR: nilearn.decomposition.tests.test_base.test_mask_reducer
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Miniconda-x64\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\Miniconda-x64\lib\site-packages\nilearn\decomposition\tests\test_base.py", line 41, in test_mask_reducer
    n_jobs=2, random_state=0)
  File "C:\Miniconda-x64\lib\site-packages\nilearn\decomposition\base.py", line 110, in mask_and_reduce
    ) for img, confound in zip(imgs, confounds))
  File "C:\Miniconda-x64\lib\site-packages\sklearn\externals\joblib\parallel.py", line 818, in __call__
    self._terminate_pool()
  File "C:\Miniconda-x64\lib\site-packages\sklearn\externals\joblib\parallel.py", line 549, in _terminate_pool
    self._pool.terminate()  # terminate does a join()
  File "C:\Miniconda-x64\lib\site-packages\sklearn\externals\joblib\pool.py", line 583, in terminate
    super(MemmapingPool, self).terminate()
  File "C:\Miniconda-x64\lib\multiprocessing\pool.py", line 465, in terminate
    self._terminate()
  File "C:\Miniconda-x64\lib\multiprocessing\util.py", line 207, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "C:\Miniconda-x64\lib\multiprocessing\pool.py", line 513, in _terminate_pool
    p.terminate()
  File "C:\Miniconda-x64\lib\multiprocessing\process.py", line 137, in terminate
    self._popen.terminate()
  File "C:\Miniconda-x64\lib\multiprocessing\forking.py", line 312, in terminate
    _subprocess.TerminateProcess(int(self._handle), TERMINATE)
WindowsError: [Error 5] Access is denied

File: examples/01_plotting/plot_demo_glass_brain.py [hunk 9]
```diff
@@ -20,10 +20,15 @@
 # demo glass brain plotting
 from nilearn import plotting
 
+# Whole brain saggital cuts
```
Comment (GaelVaroquaux) [inline examples/01_plotting/plot_demo_glass_brain.py]: To make the HTML view of the example simpler to comprehend, I think that it would be good to have 3 different "cells" in this example, one per plot.

File: nilearn/plotting/displays.py (hunk 4 above)
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: I don't really mind if the modifications propagate up if it makes the code simpler. For classes that don't accept "l" or "r", we can always capture this in the __init__

Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: Also, I forgot to mention, but I would love the connectome plots to have logic for "l" and "r" too.

File: nilearn/plotting/glass_brain.py [hunk 10]
```diff
@@ -87,12 +87,16 @@ def _get_json_and_transform(direction):
     """
     direction_to_view_name = {'x': 'side',
                               'y': 'front',
-                              'z': 'top'}
+                              'z': 'top',
+                              'l': 'left_side',
+                              'r': 'side'}
```
Comment (GaelVaroquaux) [inline nilearn/plotting/glass_brain.py]: That last one should be "right_side", no?

**Comment by chrisfilo** [comment]
BTW if you would like to check how well this performs on a diverse set of maps you can download this NeuroVault snapshot and generate figures for each map: https://purl.stanford.edu/qd938sk1220

File: nilearn/plotting/displays.py (hunk 4 above)
Comment (AlexandreAbraham) [inline nilearn/plotting/displays.py]: +1 for pushing the code in upper classes. If you are afraid of breaking stuff, maybe we can do it in 2 times? Merging this one. and refactoring afterward.

File: nilearn/plotting/displays.py [hunk 11]
```diff
@@ -350,6 +350,115 @@ def _add_lines(self, line_coords, line_values, cmap,
...
+        """
+        if self.direction in 'lr':
+            max_axis = 0
+        else:
+            max_axis = '.yz'.index(self.direction)
+
+        # set unselected brain hemisphere activations to 0
+        if self.direction == 'l':
+            data_selection = data.copy()
```
Comment (AlexandreAbraham) [inline nilearn/plotting/displays.py]: I'm pretty sure that this copy is not necessary. This is not a big deal, though.

File: nilearn/plotting/displays.py [hunk 12]
```diff
@@ -350,6 +350,115 @@ def _add_lines(self, line_coords, line_values, cmap,
...
+            data_selection[(data.shape[0]/2):, :, :] = 0
```
Comment (AlexandreAbraham) [inline nilearn/plotting/displays.py]: Assuming that data is centered seems dangerous to me. You should use the affine to get the "real" coordinate of the center.

File: nilearn/plotting/displays.py [hunk 13]
```diff
@@ -350,6 +350,115 @@ def _add_lines(self, line_coords, line_values, cmap,
...
+            inds = [a1, a2]
+            inds.insert(max_axis, np.abs(data_selection).argmax(axis=max_axis))
+
+            # take the values where the absolute value of the projection
+            # is the highest
+            maximum_intensity_data = data[inds]
+        else:
+            maximum_intensity_data = np.abs(data_selection).max(axis=max_axis)
+
+        if self.direction == 'l':
```
Comment (AlexandreAbraham) [inline nilearn/plotting/displays.py]: Can you comment on these lines?

File: nilearn/plotting/glass_brain.py (hunk 10 above)
Comment (MartinPerez) [inline nilearn/plotting/glass_brain.py]: r reuses the same image as x. So I left it with the same id when getting info from the files.

File: nilearn/plotting/displays.py (hunk 11 above)
Comment (MartinPerez) [inline nilearn/plotting/displays.py]: How should it be done? I was just trying to avoid modifying the original data "passed by reference" to the transform_to_2d function. Could I instead just do data=data to create a new function local variable?

File: nilearn/plotting/displays.py (hunk 12 above)
Comment (MartinPerez) [inline nilearn/plotting/displays.py]: I see your point, i will address this

File: nilearn/plotting/displays.py (hunk 13 above)
Comment (MartinPerez) [inline nilearn/plotting/displays.py]: ok. enough to clarify i am flipping in the saggital cut?

File: nilearn/plotting/glass_brain.py (hunk 10 above)
Comment (GaelVaroquaux) [inline nilearn/plotting/glass_brain.py]: > r reuses the same image as x. So I left it with the same id when getting info
> from the files.

Of course. Sorry for the noise.

File: nilearn/plotting/displays.py (hunk 4 above)
Comment (MartinPerez) [inline nilearn/plotting/displays.py]: I would really prefer to merge and refactor in another PR. Also I will need someone to give me a hand with the connectome since I dont work with these I am not sure I understand what it means for the connectome plots to have 'l' and 'r' logic. Would it be in the 'l' case just to overlay the same markers and lines flipped?

**Comment by MartinPerez** [comment]
@AlexandreAbraham I addressed your comments. We merge the PR and start a new one to push to upper classes and add connectomics?

**Comment by MartinPerez** [comment]
@chrisfilo thanks for the reference, I will check it out.

File: nilearn/plotting/displays.py [hunk 14]
```diff
@@ -350,6 +350,121 @@ def _add_lines(self, line_coords, line_values, cmap,
             self.ax.add_line(line)
 
 
+class GlassBrainHemisphericAxes(GlassBrainAxes):
+    """An MPL axis-like object that displays a 2D projection of 3D
+    volumes with a schematic view of the brain. In the case of saggital cuts,
+    projects only data in the selected left or right hemisphere.
+
```
Comment (banilo) [inline nilearn/plotting/displays.py]: empty line

File: nilearn/plotting/displays.py [hunk 15]
```diff
@@ -350,6 +350,121 @@ def _add_lines(self, line_coords, line_values, cmap,
...
+    """
+    def __init__(self, ax, direction, coord, plot_abs=True, **kwargs):
+        super(GlassBrainHemisphericAxes, self).__init__(
+            ax, direction, coord, plot_abs, **kwargs)
+
+    def draw_2d(self, data_2d, data_bounds, bounding_box,
+                type='imshow', **kwargs):
+        # kwargs massaging
```
Comment (banilo) [inline nilearn/plotting/displays.py]: massage or message? :-P

**Comment by GaelVaroquaux** [comment]
> We merge the PR and start a new one to push to upper classes and add connectomics?

I would like the code duplication to be addressed before this lands in
master. Code duplication makes the repo hard to read and understand. I
don't want to degrade the quality of the repo, even with the promise that
it will improve later.

**Comment by MartinPerez** [comment]
> I would like the code duplication to be addressed before this lands in master
ok I will work on the upper classes then. 

Who can give me a hand with checking the logic of the connectomics, or that I leave for later?

**Comment by MartinPerez** [comment]
connectome

well connectome included and code generalized to upper classes. I also figured out that it was completely unnecessary to have the svg image for the right hemisphere, since the matplotlib x axis could be inverted.

**Comment by MartinPerez** [comment]
of course added smoke test and example update for connectome

File: examples/01_plotting/plot_demo_glass_brain.py [hunk 16]
```diff
@@ -26,4 +26,11 @@
     localizer_tmap_filename, title='plot_glass_brain',
     black_bg=True, display_mode='xz', threshold=3)
 
+# plotting.show()
```
Comment (aabadie) [inline examples/01_plotting/plot_demo_glass_brain.py]: Maybe this comment could be removed.

File: examples/01_plotting/plot_demo_glass_brain.py [hunk 17]
```diff
@@ -17,7 +17,7 @@
 localizer_tmap_filename = localizer_dataset.tmaps[1]
 
 ###############################################################################
-# demo glass brain plotting
+# demo glass brain plotting. Whole brain saggital cuts
```
Comment (aabadie) [inline examples/01_plotting/plot_demo_glass_brain.py]: "Demo glass brain plotting using whole brain saggital cuts."

File: nilearn/plotting/displays.py [hunk 18]
```diff
@@ -237,7 +245,24 @@ def transform_to_2d(self, data, affine):
...
+            max_axis = 0
+        else:
+            max_axis = '.yz'.index(self.direction)
+
+        # set unselected brain hemisphere activations to 0
+        if self.direction == 'l':
+            data_selection = data.copy()
+            x_center, _, _, _ = np.dot(np.linalg.inv(affine),
+                                       np.array([0, 0, 0, 1]))
+            data_selection[int(x_center):, :, :] = 0
```
Comment (AlexandreAbraham) [inline nilearn/plotting/displays.py]: This will not work if the affine is not diagonal. Is that always the case? If so, you could just do `data_selection=data[:int(x_center), :, :]` to avoid the copy above.

**Comment by AlexandreAbraham** [comment]
I just have one question. When addressed (along with the remaining comment of aabadie), I will merge.

File: nilearn/plotting/displays.py (hunk 18 above)
Comment (MartinPerez) [inline nilearn/plotting/displays.py]: @AlexandreAbraham as I understood from http://nipy.org/nibabel/coordinate_systems.html, the affine is not necessarily diagonal but is invertible and its last row is 0,0,0,1. Nibabel documentation explain that [i,j,k,1] = dot(inv(M), [x,y,z,1]) so I am trusting thats the case. Since for the glass brain we are in MNI space I am assuming I can consider space coordinates 0,0,0 to ask for the actual voxel indices corresponding to the center.

File: nilearn/plotting/displays.py [hunk 19]
```diff
@@ -237,7 +245,27 @@ def transform_to_2d(self, data, affine):
...
+        elif self.direction == 'r':
+            x_center, _, _, _ = np.dot(np.linalg.inv(affine),
+                                       np.array([0, 0, 0, 1]))
+            data_selection = data[:int(x_center), :, :]
+        else:
+            data_selection = data
+
+        # We need to make sure data_selection is not empty in the x axis
+        if data_selection.shape[0] == 0:
+            data_selection = data
```
Comment (MartinPerez) [inline nilearn/plotting/displays.py]: I am not sure this is a good way to avoid a crash during smoke testing. Basically this will avoid the function crashing if the img passed has an affine that maps coordinate 0mm to a voxel index out of the array. How should handle that case. Actually raise a warning/error to the user?

**Comment by MartinPerez** [comment]
@AlexandreAbraham Its fine now, but there is a small problem if the user tried to pass an img not in mni space, which should not be the case anyway. Could you please give me your opinion on how to handle that?

**Comment by KamalakerDadi** [comment]
> but there is a small problem if the user tried to pass an img not in mni space, which should not be the case anyway. Could you please give me your opinion on how to handle that?

For me solution to this problem is to add a line saying that images should be in MNI space. We always pressure this point to users to avoid surprising results.

**Comment by GaelVaroquaux** [comment]
> For me solution to this problem is to add a line saying that images should be
> in MNI space. We always pressure this point to users to avoid surprising
> results.

+1

**Comment by KamalakerDadi** [comment]
Please also add a line about this new feature in ```whats_new.rst```

**Comment by MartinPerez** [comment]
@KamalakerDadi I created version 0.2.5 in whats_new.rst, I assume thats the next release.

**Comment by MartinPerez** [comment]
@AlexandreAbraham it seems there is an unrelated error in travis.
ERROR: nilearn.tests.test_testing.test_memory_usage

File: examples/01_plotting/plot_demo_glass_brain.py [hunk 20]
```diff
@@ -17,7 +17,7 @@
 localizer_tmap_filename = localizer_dataset.tmaps[1]
 
 ###############################################################################
-# demo glass brain plotting
+# Demo glass brain plotting using whole brain saggital cuts
```
Comment (KamalakerDadi) [inline examples/01_plotting/plot_demo_glass_brain.py]: what it means by whole brain saggital cuts ?

Comment (KamalakerDadi) [inline examples/01_plotting/plot_demo_glass_brain.py]: saggital -> sagittal

File: examples/01_plotting/plot_demo_glass_brain.py [hunk 21]
```diff
@@ -26,4 +26,9 @@
     localizer_tmap_filename, title='plot_glass_brain',
     black_bg=True, display_mode='xz', threshold=3)
 
+###############################################################################
+# Hemispheric saggital cuts
+plotting.plot_glass_brain(
+    localizer_tmap_filename, display_mode='lyrz', threshold=3)
```
Comment (KamalakerDadi) [inline examples/01_plotting/plot_demo_glass_brain.py]: It would be easy to catch quickly using title something like plot_glass_brain with display_mode='lyrz'

File: examples/01_plotting/plot_demo_glass_brain_extensive.py [hunk 22]
```diff
@@ -36,3 +37,25 @@
                           colorbar=True, plot_abs=False)
 
 plotting.show()
+
+###############################################################################
+# Hemispheric saggital cuts
+plotting.plot_glass_brain(localizer_tmap_filename, threshold=3,
+                          display_mode='lyrz')
```
Comment (KamalakerDadi) [inline examples/01_plotting/plot_demo_glass_brain_extensive.py]: same with here some ```title```s would be useful.

**Comment by MartinPerez** [comment]
took care of all comments and tests passed except for the unrelated memory profiler one in travis.

File: doc/whats_new.rst [hunk 23]
```diff
@@ -1,3 +1,14 @@
+0.2.5
+=====
+
+Changelog
+---------
+
+New features
+
```
Comment (lesteve) [inline doc/whats_new.rst]: This needs to be a header, look at how it is done for the 0.2.4 section.

File: examples/01_plotting/plot_demo_glass_brain_extensive.py [hunk 24]
```diff
@@ -36,3 +37,31 @@
...
+
+title = 'plot_glass_brain with display_mode="lyrz"'
+plotting.plot_glass_brain(localizer_tmap_filename, threshold=0, colorbar=True,
+                          title=title, plot_abs=False, display_mode='lyrz')
+
+title = 'plot_glass_brain with display_mode="lyrz"'
+plotting.plot_glass_brain(localizer_tmap_filename, threshold=3, title=title,
+                          colorbar=True, plot_abs=False, display_mode='lyrz')
+
+plotting.show()
```
Comment (lesteve) [inline examples/01_plotting/plot_demo_glass_brain_extensive.py]: Dow we really need all this variations?

File: nilearn/plotting/displays.py [hunk 25]
```diff
@@ -266,9 +295,23 @@ def draw_position(self, size, bg_color, **kwargs):
...
-
         xdata, ydata = marker_coords_2d.T
 
+        # Allow markers only in their respective hemisphere when appropriate
+        if self.direction in 'lr':
+            relevant_coords = []
+            xcoords, ycoords, zcoords = marker_coords.T
+            for cidx, xc in enumerate(xcoords):
+                if self.direction == 'r':
+                    if xc >= 0:
```
Comment (lesteve) [inline nilearn/plotting/displays.py]: you can put that in the same if clause: `if self.direction == 'r' and xc >=0`

File: nilearn/plotting/displays.py [hunk 26]
```diff
@@ -266,9 +295,23 @@ def draw_position(self, size, bg_color, **kwargs):
...
+                        relevant_coords.append(cidx)
+                elif self.direction == 'l':
+                    if xc < 0:
```
Comment (lesteve) [inline nilearn/plotting/displays.py]: same remark as above you can put it in the same if clause.

File: nilearn/plotting/img_plotting.py [hunk 27]
```diff
@@ -957,20 +957,26 @@ def plot_glass_brain(stat_map_img,
...
+        properly.
+
         Parameters
         ----------
         stat_map_img : Niimg-like object
             See http://nilearn.github.io/manipulating_images/manipulating_images.html#niimg.
             The statistical map image. It needs to be in MNI space
             in order to align with the brain schematics.
-        output_file : string, or None, optional
+        output_file : string, or None, optional.
```
Comment (lesteve) [inline nilearn/plotting/img_plotting.py]: I wouldn't add a dot at the end since that's not what is done for other parameters

**Comment by lesteve** [comment]
It would be worth to modify the docs to show the new modes in http://nilearn.github.io/plotting/index.html#different-display-modes.

File: examples/01_plotting/plot_demo_glass_brain_extensive.py (hunk 24 above)
Comment (AlexandreAbraham) [inline examples/01_plotting/plot_demo_glass_brain_extensive.py]: +1, I guess 2 plots are enough ;).

File: nilearn/plotting/displays.py [hunk 28]
```diff
@@ -237,7 +245,28 @@ def transform_to_2d(self, data, affine):
...
+        if self.direction == 'l':
+            x_center, _, _, _ = np.dot(np.linalg.inv(affine),
+                                       np.array([0, 0, 0, 1]))
+            data_selection = data[int(x_center):, :, :]
+        elif self.direction == 'r':
+            x_center, _, _, _ = np.dot(np.linalg.inv(affine),
+                                       np.array([0, 0, 0, 1]))
+            data_selection = data[:int(x_center), :, :]
+        else:
+            data_selection = data
```
Comment (AlexandreAbraham) [inline nilearn/plotting/displays.py]: The code above will only work if the affine is diagonal. Are we sure that it is diagonal at this point?

Comment (MartinPerez) [inline nilearn/plotting/displays.py]: @AlexandreAbraham as I understood from http://nipy.org/nibabel/coordinate_systems.html, the affine is not necessarily diagonal but is invertible and its last row is 0,0,0,1. Nibabel documentation explain that [i,j,k,1] = dot(inv(M), [x,y,z,1]) so I am trusting thats the case. Since for the glass brain we are in MNI space I am assuming I can consider space coordinates 0,0,0 to ask for the actual voxel indices corresponding to the center.

Does this answer your question?

**Comment by GaelVaroquaux** [comment]
What's the status on the hemisphere plotting of connectomes in this PR?

PS: sorry, I think that I just introduced a trivial conflict with master.

**Comment by MartinPerez** [comment]
@GaelVaroquaux its also ready for the connectome. Look for the image above.

**Comment by GaelVaroquaux** [comment]
Indeed, very nice!

We would need a tiny example demoing it.

Maybe you can do a new example in the plotting section that demos plot connectome and replace the image in https://circle-artifacts.com/gh/nilearn/nilearn/1495/artifacts/0/home/ubuntu/nilearn/doc/_build/html/plotting/index.html#different-plotting-functions by the image it generates (and link to that example).

I think that, for now, we could use a hard-coded set of coordinates and connectivity matrix, for instance inspired by: https://circle-artifacts.com/gh/nilearn/nilearn/1495/artifacts/0/home/ubuntu/nilearn/doc/_build/html/auto_examples/03_connectivity/plot_adhd_spheres.html#sphx-glr-auto-examples-03-connectivity-plot-adhd-spheres-py

It would enable us to demo the hemisphere plotting for connectomes.

File: nilearn/plotting/displays.py (hunk 26 above)
Comment (lesteve) [inline nilearn/plotting/displays.py]: By the way just wondering what should we do if xc == 0 (maybe that never happens in practice I am not sure). Should we not show the marker on both 'l' and 'r' plots?

Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: > By the way just wondering what should we do if xc == 0 (maybe that never
> happens in practice I am not sure). Should we not show the marker on both 'l'
> and 'r' plots?

Actually, that's probably a good idea. And that way we have edges going
to it in both plot (this needs to be documented in the docstrings of
add_marker and plot_connectome).

**Comment by KamalakerDadi** [comment]
>Should I change the folder name or is this a new standard to enforce in all links?

No, you don't have to it will generate it automatically while you ```make html``` in doc folder.

**Comment by MartinPerez** [comment]
@lesteve I illustrate the new display modes using two examples from glass brain. Did not make it extensive in the images but instead added reference to the remaining possible modes in the text at the side.

@GaelVaroquaux I modified the simple example you mentioned and emphasized the case x==0.
https://circle-artifacts.com/gh/nilearn/nilearn/1520/artifacts/0/home/ubuntu/nilearn/doc/_build/html/auto_examples/03_connectivity/plot_adhd_spheres.html

Also made link to image and examples in https://circle-artifacts.com/gh/nilearn/nilearn/1520/artifacts/0/home/ubuntu/nilearn/doc/_build/html/plotting/index.html#different-display-modes

Also addressed remaining comments.

**Comment by bthirion** [comment]
LGTM.

**Comment by AlexandreAbraham** [comment]
LGTM too. Merging. Thanks a lot!

**Comment by GaelVaroquaux** [comment]
Whoooot

**Comment by MartinPerez** [comment]
Awesome! thanks!
